File: code/backend/src/main/java/com/mysterytrip/controller/BusController.java
package com.mysterytrip.controller;

import com.mysterytrip.service.PublicBusService;
import com.mysterytrip.service.PublicBusService.Arrival;
import com.mysterytrip.service.PublicBusService.Station;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

/** TAGO 전국 버스 도착정보 프록시 — 프론트 지도에서 실시간 도착 안내용 */
@RestController
@RequestMapping("/api/bus")
public class BusController {

    private final PublicBusService busService;

    public BusController(PublicBusService busService) {
        this.busService = busService;
    }

    /**
     * GET /api/bus/arrivals?lat=..&lng=..&busNo=10
     * 좌표에서 가장 가까운 정류소의 실시간 버스 도착 예정 (busNo 지정 시 해당 노선만)
     */
    @GetMapping("/arrivals")
    public Map<String, Object> arrivals(@RequestParam double lat,
                                        @RequestParam double lng,
                                        @RequestParam(required = false) String busNo) {
        if (!busService.enabled()) {
            return Map.of("enabled", false, "arrivals", List.of());
        }
        Map.Entry<Station, List<Arrival>> result = busService.arrivalsNear(lat, lng, busNo);
        if (result == null) {
            return Map.of("enabled", true, "arrivals", List.of());
        }
        return Map.of(
                "enabled", true,
                "station", Map.of("name", result.getKey().name(), "nodeId", result.getKey().nodeId()),
                "arrivals", result.getValue()
        );
    }
}

File: code/backend-kotlin/src/main/kotlin/com/ttarawayou/controller/BusController.kt
package com.ttarawayou.controller

import com.ttarawayou.service.PublicBusService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/** TAGO 전국 버스 도착정보 프록시 — 프론트 지도에서 실시간 도착 안내용 */
@RestController
@RequestMapping("/api/bus")
class BusController(private val busService: PublicBusService) {

    /**
     * GET /api/bus/arrivals?lat=..&lng=..&busNo=10
     * 좌표에서 가장 가까운 정류소의 실시간 버스 도착 예정 (busNo 지정 시 해당 노선만)
     */
    @GetMapping("/arrivals")
    fun arrivals(
        @RequestParam lat: Double,
        @RequestParam lng: Double,
        @RequestParam(required = false) busNo: String?,
    ): Map<String, Any> {
        if (!busService.enabled()) {
            return mapOf("enabled" to false, "arrivals" to emptyList<Any>())
        }
        val result = busService.arrivalsNear(lat, lng, busNo)
            ?: return mapOf("enabled" to true, "arrivals" to emptyList<Any>())
        return mapOf(
            "enabled" to true,
            "station" to mapOf("name" to result.first.name, "nodeId" to result.first.nodeId),
            "arrivals" to result.second,
        )
    }
}
